### ISSUE TITLE
Possibility to provide an instance of WebSocket, instead of URI to RFB

### ISSUE THREAD

**Issue by moltar** [novnc/noVNC#1243]
**Is your feature request related to a problem? Please describe.**

It would be useful to be able to provide a `WebSocket` object, instead of a URI to the `RFB` object.

This would allow for better control of WebSocket connections. E.g. attaching error handles to alert a user of various errors.

**Describe the solution you'd like**

```js
const ws = new WebSocket('ws://...')
const rfb = new RFB(el, ws)
```

**Describe alternatives you've considered**

Exposing `WebSocket` object from the `RFB` object would be a viable alternative too. But worse software-design wise.

**Additional context**

N/A

**Comment by CendioOssman** [comment]
That sounds very reasonable. Do you feel like writing a suggested patch?

**Comment by moltar** [comment]
I could try.

Is there a contribution guide?

As I see this project doesn't follow common NPM package structure.

Any tips on where to start?

**Comment by CendioOssman** [comment]
I'm afraid we do not have any guide. We don't really pay that much attention to NPM since we require a browser to function. We're only on there since people use NPM as a convenient way to package up their things.

I think most of this is in the `RFB` constructor (in `core/rfb.js`), so start there. You might also want to switch over to using the `addEventListener()` API to avoid conflicts with listeners the caller might have set up.

**Comment by moltar** [comment]
> You might also want to switch over to using the `addEventListener()` API to avoid conflicts with listeners the caller might have set up.

Could you please elaborate on that? I am not sure how this fits in with my requirement?

Is there an alternative `RFB` API available?

I was previously following this documentation: https://github.com/novnc/noVNC/blob/master/docs/API.md

**Comment by CendioOssman** [comment]
I meant the API for the WebSocket object. Right now we are using things like `onerror` instead of `addEventListener('error', ...)`, which isn't ideal.

**Comment by moltar** [comment]
Oh, right, I see what you mean. Basically, right now it is setting an event listener via prop assignment, which then means we can only have one event listener. And if we refactor to use `addEventListener`, then we could attach multiple listeners per WS instance.

**Comment by CendioOssman** [comment]
This got fixed in #1523.
